### PR TITLE
XWIKI-6732 : Introduce DocumentMovingEvent and DocumentMovedEvent events

### PR DESCRIPTION
**Pull request by SrijanJha28**

### CODE REVIEW

File: xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+ * <p>
+ * The event also send the following parameters:
+ * </p>
+ * <ul>
+ * <li>source: the current {com.xpn.xwiki.doc.XWikiDocument} instance</li>
+ * <li>data: the current {com.xpn.xwiki.XWikiContext} instance</li>
+ * </ul>
+ * 
+ * @version $Id$
+ * @since 9.1RC1
```
Comment (vmassol) [inline xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java]: Wrong Since :)

File: xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+ */
+public class DocumentMovedEvent extends AbstractDocumentEvent
```
Comment (vmassol) [inline xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java]: Would be good to add a `@Unstable` annotation too. (same below for the other class too)

File: xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java (hunk 1 above)
Comment (SrijanJha28) [reply xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java]: what should be the since version?

File: xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java (hunk 2 above)
Comment (SrijanJha28) [reply xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java]: ok

File: xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    /**
+     * The version identifier for this Serializable class. Increment only if the <i>serialized</i> form of the class
+     * changes.
+     */
+    private static final long serialVersionUID = 1L;
+    
+    /**
+     * Constructor initializing the event filter with an
+     * {@link org.xwiki.observation.event.filter.AlwaysMatchingEventFilter}, meaning that this event will match any
+     * other document delete event.
```
Comment (vmassol) [inline xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java]: wrong copy paste?

File: xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+     */
+    public DocumentMovedEvent()
+    {
+        super();
+    }
+    
+    /**
+     * Constructor initializing the event filter with a {@link org.xwiki.observation.event.filter.FixedNameEventFilter},
+     * meaning that this event will match only delete events affecting the same document.
```
Comment (vmassol) [inline xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java]: Wrong copy paste?

File: xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java (hunk 1 above)
Comment (vmassol) [reply xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java]: `@since 10.0RC1`

File: xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java (hunk 3 above)
Comment (SrijanJha28) [reply xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java]: what's wrong in pasting?

File: xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java (hunk 4 above)
Comment (SrijanJha28) [reply xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java]: what's wrong in pasting?

Comment (vmassol) [reply xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java]: Pasting is not the goal. The goal is to have some meaningful javadoc! :)

Comment (vmassol) [reply xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java]: Apparently you haven't even read what you pasted... ;)

**Comment by SrijanJha28** [comment]
Done :)

**Comment by vmassol** [comment]
Thx. Javadoc are still not correct. Please read what you pasted there and fix them please.

File: xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java [hunk 5]
```diff
@@ -44,25 +44,42 @@
...
+     */
+    public DocumentMovedEvent(DocumentReference documentReference)
+    {
+        this(documentReference, null);
+    }
+
     /**
-     * Constructor initializing the event filter with a {@link org.xwiki.observation.event.filter.FixedNameEventFilter}, meaning that this event will match only delete events affecting the same document.
+     * Matches {@link DocumentMovedEvent} events that target the specified document and move. The move is
+     * matched only if it's not {@code null}.
```
Comment (vmassol) [inline xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java]: Why did you remove the explanation about the constructor's initializang the event filter etc?

All you needed was to remove the part about "delete" which is not correct since this is a "move"...

File: xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java [hunk 6]
```diff
@@ -44,25 +44,42 @@
...
      * 
      * @param documentReference the reference of the document to match
+     * @param move the move the document was moved to
      */
-    public DocumentMovedEvent(DocumentReference documentReference)    
+    public DocumentMovedEvent(DocumentReference documentReference, String move)
```
Comment (vmassol) [inline xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java]: I think the moved document should be contained in the data part instead.

Comment (SrijanJha28) [reply xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java]: But this also works fine for me

File: xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java (hunk 5 above)
Comment (SrijanJha28) [reply xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java]: should I also add the constructor's initializang the event filter

**Comment by tmortagne** [comment]
You also need to send those events (otherwise what the javadoc say is false since they are never sent).

File: xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+        if (matches) {
+            DocumentMovedEvent documentMovedEvent = (DocumentMovedEvent) otherEvent;
+            matches = move == null || move.equals(documentMovedEvent.getMove());
+        }
+
+        return matches;
+    }
+
+private void move(DocumentReference documentReference, EventFilter eventFilter)
```
Comment (tmortagne) [inline xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java]: I don't understand what this method is doing here.

Comment (SrijanJha28) [reply xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java]: Should i remove this method?

File: xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+ * <ul>
+ * <li>source: the current {com.xpn.xwiki.doc.XWikiDocument} instance</li>
+ * <li>data: the current {com.xpn.xwiki.XWikiContext} instance</li>
+ * </ul>
+ * 
+ * @version $Id$
+ * @since 10.0RC1
+ */
+@Unstable 
+public class DocumentMovedEvent extends AbstractDocumentEvent
```
Comment (tmortagne) [inline xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java]: This even should implement `org.xwiki.observation.event.EndEvent`.

File: xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovingEvent.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+ * <ul>
+ * <li>source: the current {com.xpn.xwiki.doc.XWikiDocument} instance</li>
+ * <li>data: the current {com.xpn.xwiki.XWikiContext} instance</li>
+ * </ul>
+ * 
+ * @version $Id$
+ * @since 10.0RC1
+ */
+@Unstable 
+public class DocumentMovingEvent extends AbstractDocumentEvent
```
Comment (tmortagne) [inline xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovingEvent.java]: This even should implement `org.xwiki.observation.event.BeginEvent`.

File: xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovingEvent.java [hunk 10]
```diff
@@ -21,6 +21,7 @@
 
 import org.xwiki.model.reference.DocumentReference;
 import org.xwiki.observation.event.filter.EventFilter;
+import org.xwiki.observation.event.BeginEvent;
```
Comment (tmortagne) [inline xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovingEvent.java]: This is just an import, you need to make the class implement it too.

**Comment by SrijanJha28** [comment]
Is the DocumentMovedEvent correct?

File: xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+     /**
+     * Matches {@link DocumentMovedEvent} events that target the specified document and move. The move is
+     * matched only if it's not {@code null}.
+     * 
+     * @param documentReference the reference of the document to match
+     * @param move the move the document was moved to
+     */
+    public DocumentMovedEvent(DocumentReference documentReference)
+    {
+        String move;
```
Comment (tmortagne) [inline xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java]: This does not build.

File: xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+     */
+    private static final long serialVersionUID = 1L;
+    
+   /**
+     * Constructor initializing the event filter with an
+     * {@link org.xwiki.observation.event.filter.AlwaysMatchingEventFilter}, meaning that this event will match any other document move event.
+     */
+     
+    /**
+     * The move document is about to be moved to.
```
Comment (tmortagne) [inline xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/event/DocumentMovedEvent.java]: Does not really mean anything.

**Comment by SrijanJha28** [comment]
Done :)
